src/main/java/com/glencoesoftware/omero/ms/image/region/HistogramRequestHandler.java

```diff
@@ -0,0 +1,300 @@
+
+    /**
+     * Constructor. Populates histogramCtx and pixelsService
+     * @param histogramCtx
+     * @param pixelsService
+     */
+    public HistogramRequestHandler(HistogramCtx histogramCtx,
+            PixelsService pixelsService) {
+        this.histogramCtx = histogramCtx;
+        this.pixelsService = pixelsService;
+    }
+
+    /**
+     * Get the min and max for the channel from the StatsInfo
+     * @param channel The channel to get the StatsInfo min and max for
+     * @return double array [min, max] if the StatsInfo is present, else null
+     */
+    private double[] getMinMaxFromStatsinfo(Channel channel) {
+        double min, max;
+        if (channel.getStatsInfo() != null) {
+            min = channel.getStatsInfo().getGlobalMin();
+            max = channel.getStatsInfo().getGlobalMax();
+            // if max == 1.0 the global min/max probably has not been
+            // calculated; fall back to plane min/max
+            if (max != 1.0)
+                return new double[] { min, max };
+        }
+        return null;
+    }
+
+    /**
+     * Get the minimum and maximum value of the pixel data
+     *
+     * @param pd
+     *            The {@link PixelData}
+     * @param channel
+     *            The {@link Channel}
+     * @return See above
+     */
+    private double[] getMinMaxFromPixelData(PixelData pd, Channel channel) {
+        double min, max;
+
+        StatsFactory sf = new StatsFactory();
+        double[] pixelMinMax = sf.initPixelsRange(channel.getPixels());
+
+        min = pixelMinMax[1];
+        max = pixelMinMax[0];
+
+        for (int i = 0; i < pd.size(); i++) {
+            min = Math.min(min, pd.getPixelValue(i));
+            max = Math.max(max, pd.getPixelValue(i));
+        }
+
+        return new double[] { min, max };
+    }
+
+    /**
+     * Read through the pixel data and produce histogram data reflecting
+     * the occurrance of pixel values within each bin
+     * @param pd The {@link PixelData} to get the pixel values from
+     * @param channel The {@link Channel} to get the pixel values from
+     * @param minMax The min and max to values to divide into bins
+     * @param imgWidth The sizeX of the pixel data
+     * (not necessarily full resolution)
+     * @param imgHeight The sizeY of the pixel data
+     * (not necessarily full resolution)
+     * @return {@link JsonArray} containing histogram data
+     */
+    public JsonArray getHistogramData(PixelData pd,
+            double[] minMax, int imgWidth, int imgHeight) {
+        int[] counts = new int[histogramCtx.bins];
+
+        double min = minMax[0];
+        double max = minMax[1];
+
+        double range = max - min + 1;
+        double binRange = range / histogramCtx.bins;
+        for (int i = 0; i < pd.size(); i++) {
+            int pdx = i % imgWidth;
+            int pdy = i / imgWidth;
+            if (pdx >= 0 && pdx < (0 + imgWidth) && pdy >= 0 && pdy < (0 + imgHeight)) {
```

Comment:
In the case of this first implementation, it looks like this check could be simplified to `i >=0 && i < (imgWidth * imgHeight)` or even moved to limit the maximum value of the for loop.

In the server-side code, the reason for having this logic is histograms can be requested for a given region specified by `x,y,w,h`. Is that a feature we might be aiming to support at some point?

Reply:
I could see us supporting this at some point which is why I left it in, but really since we're only doing full images none of these checks are necessary right now. I'm happy to take them out if you think they're just cluttering things up.